File: backend/ai_service.py
import logging
from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_core.prompts import PromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
import os
from dotenv import load_dotenv
import google.generativeai as genai

# Configure logging
logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)

# Directly implement the API key
api_key = "AIzaSyC7958lvH9cC0y4E2qUV26o-0eBmDB1L70"  # Replace with your actual API key

# Attempt to load from .env file as fallback
if not api_key or api_key == "YOUR_ACTUAL_API_KEY_HERE":
    load_dotenv()
    api_key = os.getenv("GOOGLE_API_KEY")
    if not api_key:
        raise ValueError("No GOOGLE_API_KEY found. Please add it directly in the code or in your .env file.")

try:
    # Initialize the direct API first
    genai.configure(api_key=api_key)
    
    # List available models first
    print("Listing available models...")
    try:
        models = genai.list_models()
        print("Available models:")
        for model in models:
            print(f" - {model.name}: {model.display_name}")
            supported_methods = []
            if hasattr(model, 'supported_generation_methods'):
                supported_methods = model.supported_generation_methods
            print(f"   Supported methods: {supported_methods}")
    except Exception as list_error:
        print(f"Error listing models: {list_error}")
    
    # Try a different model name - gemini-1.5-pro is the newer version
    print("Trying with gemini-1.5-pro model...")
    try:
        model = genai.GenerativeModel('gemini-1.5-pro')
        test_response = model.generate_content("Test")
        print(f"Direct API test succeeded with gemini-1.5-pro: {test_response.text}")
    except Exception as pro_error:
        print(f"Error with gemini-1.5-pro: {pro_error}")
        
        # Try with gemini-1.5-flash as a fallback
        print("Trying with gemini-1.5-flash model...")
        try:
            model = genai.GenerativeModel('gemini-1.5-flash')
            test_response = model.generate_content("Test")
            print(f"Direct API test succeeded with gemini-1.5-flash: {test_response.text}")
            
            # Use the working model name
            working_model = 'gemini-1.5-flash'
        except Exception as flash_error:
            print(f"Error with gemini-1.5-flash: {flash_error}")
            working_model = None
    else:
        # gemini-1.5-pro worked
        working_model = 'gemini-1.5-pro'
    
    if working_model:
        print(f"Using {working_model} for LangChain integration")
        # Initialize LangChain with the working model
        llm = ChatGoogleGenerativeAI(
            model=working_model,
            google_api_key=api_key,
            temperature=0.2,
            top_p=0.8,
            top_k=40,
            verbose=True
        )
        output_parser = StrOutputParser()
        print("LLM initialized successfully")
    else:
        print("ERROR: Could not find a working Gemini model!")
        llm = None
        output_parser = None
except Exception as init_error:
    print(f"ERROR initializing AI: {str(init_error)}")
    import traceback
    print(traceback.format_exc())
    llm = None
    output_parser = None

app = FastAPI()

# Configure CORS
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],  # Use ["http://localhost:5173"] for more security
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

class AIRequest(BaseModel):
    code: str
    language: str
    prompt: str

class AIResponse(BaseModel):
    updated_code: str
    explanation: str

# Update the /ai/improve endpoint
@app.post("/ai/improve", response_model=AIResponse)
async def improve_code(request: AIRequest):
    if llm is None:
        raise HTTPException(
            status_code=503,
            detail="AI service unavailable: No working Gemini model found"
        )
    try:
        print(f"Received request: language={request.language}, prompt length={len(request.prompt)}, code length={len(request.code)}")
        
        # Create a prompt template
        template = """
        You are an expert programmer focusing on improving code.
        
        ORIGINAL CODE ({language}):
        ```{language}
        {code}
        ```
        
        USER REQUEST:
        {prompt}
        
        Please improve the code according to the user's request. 
        Follow these rules:
        1. Maintain the same general functionality unless the user explicitly asks for changes
        2. Make the code more efficient, readable, or maintainable as appropriate
        3. Add comments to explain complex parts
        4. Format your response in two parts: 
           - First provide ONLY the improved code without markdown formatting or explanation
           - Then on a new line write "EXPLANATION:" followed by your explanation
        
        IMPROVED CODE:
        """
        
        prompt_template = PromptTemplate(
            input_variables=["language", "code", "prompt"],
            template=template
        )
        
        print("Creating chain...")
        # Create and run the chain with detailed error handling
        try:
            # Make sure output_parser is initialized
            if output_parser is None:
                output_parser = StrOutputParser()
                
            chain = (
                {"language": lambda x: x["language"], "code": lambda x: x["code"], "prompt": lambda x: x["prompt"]} 
                | prompt_template 
                | llm 
                | output_parser
            )
            
            print("Invoking chain...")
            result = chain.invoke({
                "language": request.language,
                "code": request.code,
                "prompt": request.prompt
            })
            print(f"Chain result length: {len(result)}")
            
        except Exception as chain_error:
            print(f"Chain execution error: {str(chain_error)}")
            import traceback
            print(traceback.format_exc())
            raise HTTPException(
                status_code=500, 
                detail=f"AI model error: {str(chain_error)}"
            )
        
        # Parse the result to separate code and explanation
        parts = result.split("EXPLANATION:", 1)
        if len(parts) > 1:
            updated_code = parts[0].strip()
            explanation = parts[1].strip()
        else:
            # If no EXPLANATION section, try to guess the separation
            lines = result.split('\n')
            code_end_index = -1
            
            # Look for markdown code blocks or natural language transitions
            for i, line in enumerate(lines):
                if "```" in line or "Here's why:" in line or "Explanation:" in line:
                    code_end_index = i
                    break
            
            if code_end_index > 0:
                updated_code = '\n'.join(lines[:code_end_index]).strip()
                explanation = '\n'.join(lines[code_end_index:]).strip()
            else:
                # Fallback: assume 80% is code, rest is explanation
                split_point = int(len(lines) * 0.8)
                updated_code = '\n'.join(lines[:split_point]).strip()
                explanation = '\n'.join(lines[split_point:]).strip()
        
        # Clean up the code (remove markdown code blocks if present)
        if updated_code.startswith("```"):
            code_lines = updated_code.split('\n')
            # Remove first line with ``` and last line with ```
            if len(code_lines) >= 3 and "```" in code_lines[0] and "```" in code_lines[-1]:
                updated_code = '\n'.join(code_lines[1:-1])
        
        print(f"Successfully processed code improvement request")
        return AIResponse(
            updated_code=updated_code,
            explanation=explanation
        )
    
    except Exception as e:
        import traceback
        traceback_str = traceback.format_exc()
        error_msg = f"AI processing error: {str(e)}"
        print(f"ERROR: {error_msg}\n{traceback_str}")
        raise HTTPException(status_code=500, detail=error_msg)

@app.post("/ai/simple-improve")
async def simple_improve(request: AIRequest):
    try:
        # Test with direct API call - no chains
        prompt = f"""
        Improve this {request.language} code:
        ```{request.language}
        {request.code}
        ```
        
        Request: {request.prompt}
        
        Respond with only improved code:
        """
        
        response = llm.invoke(prompt)
        return {"result": str(response)}
    except Exception as e:
        import traceback
        error_msg = f"Simple AI error: {str(e)}"
        print(f"ERROR: {error_msg}\n{traceback.format_exc()}")
        return {"error": error_msg}

@app.post("/ai/direct-improve", response_model=AIResponse)
async def direct_improve(request: AIRequest):
    try:
        print(f"Direct improve: language={request.language}, prompt={request.prompt}")
        
        # Use the Google API directly without LangChain
        prompt = f"""
        You are an expert programmer focusing on improving code.
        
        ORIGINAL CODE ({request.language}):
        ```{request.language}
        {request.code}
        ```
        
        USER REQUEST:
        {request.prompt}
        
        Please improve the code according to the user's request. 
        Follow these rules:
        1. Maintain the same general functionality unless the user explicitly asks for changes
        2. Make the code more efficient, readable, or maintainable as appropriate
        3. Add comments to explain complex parts
        4. Format your response in two parts: 
           - First provide ONLY the improved code without markdown formatting or explanation
           - Then on a new line write "EXPLANATION:" followed by your explanation
        
        IMPROVED CODE:
        """
        
        # Use the model that's known to work
        model_name = 'gemini-1.5-pro'  # We know this works from your tests
        print(f"Using model: {model_name}")
        
        model = genai.GenerativeModel(
            model_name,
            generation_config=genai.types.GenerationConfig(
                temperature=0.2,
                top_p=0.8,
                top_k=40,
            )
        )
        
        # Generate the content
        response = model.generate_content(prompt)
        result = response.text
        print(f"Direct API response: {result[:100]}...")
        
        # Parse the result to separate code and explanation
        parts = result.split("EXPLANATION:", 1)
        if len(parts) > 1:
            updated_code = parts[0].strip()
            explanation = parts[1].strip()
        else:
            # If no EXPLANATION section, try to guess the separation
            lines = result.split('\n')
            code_end_index = -1
            
            # Look for markdown code blocks or natural language transitions
            for i, line in enumerate(lines):
                if "```" in line or "Here's why:" in line or "Explanation:" in line:
                    code_end_index = i
                    break
            
            if code_end_index > 0:
                updated_code = '\n'.join(lines[:code_end_index]).strip()
                explanation = '\n'.join(lines[code_end_index:]).strip()
            else:
                # Fallback: assume 80% is code, rest is explanation
                split_point = int(len(lines) * 0.8)
                updated_code = '\n'.join(lines[:split_point]).strip()
                explanation = '\n'.join(lines[split_point:]).strip()
        
        # Clean up the code (remove markdown code blocks if present)
        if updated_code.startswith("```"):
            code_lines = updated_code.split('\n')
            # Remove first line with ``` and last line with ```
            if len(code_lines) >= 3 and "```" in code_lines[0] and "```" in code_lines[-1]:
                updated_code = '\n'.join(code_lines[1:-1])
        
        print(f"Successfully processed direct code improvement request")
        return AIResponse(
            updated_code=updated_code,
            explanation=explanation
        )
    
    except Exception as e:
        import traceback
        traceback_str = traceback.format_exc()
        error_msg = f"Direct AI processing error: {str(e)}"
        print(f"ERROR: {error_msg}\n{traceback_str}")
        raise HTTPException(status_code=500, detail=error_msg)

# Add a test endpoint to check the API key
@app.get("/test-gemini")  # This must be exactly as shown here - check for typos
async def test_gemini():
    logger.debug("test-gemini endpoint called")
    try:
        # Test with a simple prompt
        logger.debug("Invoking Gemini LLM")
        simple_result = llm.invoke("Say hello")
        logger.debug(f"Received response: {simple_result}")
        return {
            "status": "success", 
            "message": "Gemini API is working", 
            "response": str(simple_result)
        }
    except Exception as e:
        logger.error(f"Error invoking Gemini: {str(e)}")
        import traceback
        logger.error(traceback.format_exc())
        return {
            "status": "error", 
            "message": f"Failed to connect to Gemini API: {str(e)}",
            "traceback": traceback.format_exc()
        }

@app.get("/direct-test")
async def direct_test():
    try:
        # Use Google's API directly
        model = genai.GenerativeModel('gemini-pro')
        response = model.generate_content("Say hello")
        return {
            "status": "success",
            "response": response.text
        }
    except Exception as e:
        import traceback
        return {
            "status": "error",
            "message": str(e),
            "traceback": traceback.format_exc()
        }

@app.get("/ping")
async def ping():
    return {"message": "pong"}

if __name__ == "__main__":
    import uvicorn
    import argparse
    
    parser = argparse.ArgumentParser()
    parser.add_argument('--port', type=int, default=8001)
    args = parser.parse_args()
    
    uvicorn.run(app, host="0.0.0.0", port=args.port)

File: backend/ai_service.js
// AI Service for code improvement using Groq API
import express from 'express';
import cors from 'cors';
import { v4 as uuidv4 } from 'uuid';
import Groq from 'groq-sdk';
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';
import http from 'http';

// Get current file directory with ES modules
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load environment variables
dotenv.config({ path: path.resolve(__dirname, '.env') });

// Configure logging
const logger = {
  debug: console.debug,
  info: console.log,
  error: console.error
};

// Use Groq API key
const groqApiKey = process.env.GROQ_API_KEY;
if (!groqApiKey) {
  throw new Error("No GROQ_API_KEY found in environment variables. Please add it to your .env file.");
}

// Initialize the Groq client
const groqClient = new Groq({ apiKey: groqApiKey });

// Select the model to use
const MODEL_NAME = "llama3-70b-8192";  // Groq's Llama 3 70B model - high quality, fast inferences

// Create a global variable to store the server reference
let server = null;
let keepAliveTimer = null;

// Test the API connection
async function testGroqApi() {
  try {
    console.log("Testing Groq API connection...");
    const response = await groqClient.chat.completions.create({
      model: MODEL_NAME,
      messages: [
        { role: "system", content: "You are a helpful assistant." },
        { role: "user", content: "Say hello" }
      ],
      temperature: 0.2,
      max_tokens: 100
    });
    
    console.log(`Groq API test succeeded: ${response.choices[0].message.content}`);
    if (response.usage.completion_time) {
      console.log(`Response completion time: ${response.usage.completion_time}s`);
    }
    return true;
  } catch (e) {
    console.error(`ERROR initializing Groq API: ${e.message}`);
    console.error(e.stack);
    throw new Error(`Failed to initialize Groq API: ${e.message}`);
  }
}

// Function to keep the server alive
function startKeepAlive() {
  // Don't create a new timer if one already exists
  if (keepAliveTimer) return;
  
  // Print a message that the server is still running every minute
  keepAliveTimer = setInterval(() => {
    console.log(`[${new Date().toLocaleTimeString()}] AI service is still running...`);
    
    // Check if we should clean up sessions
    const now = new Date();
    if ((now - LAST_CLEANUP_TIME) > SESSION_CLEANUP_INTERVAL) {
      cleanupExpiredSessions();
      LAST_CLEANUP_TIME = now;
    }
  }, 60 * 1000); // Every minute
  
  console.log("Keep-alive timer started to maintain server process");
}

// Create the express app
const app = express();

// Configure CORS
app.use(cors({
  origin: "*",
  credentials: true,
  methods: ["*"],
  allowedHeaders: ["*"]
}));

// Parse JSON request bodies
app.use(express.json());

// Define in-memory session storage (for production, use a database)
const sessions = {};

// Session cleanup interval (in milliseconds)
const SESSION_CLEANUP_INTERVAL = 3600 * 1000;  // 1 hour
let LAST_CLEANUP_TIME = new Date();
const SESSION_EXPIRY = 24 * 60 * 60 * 1000;  // 24 hours of inactivity

// Add process event listeners for graceful shutdown
process.on('SIGINT', () => {
  console.log('Received SIGINT. Shutting down AI service gracefully...');
  if (keepAliveTimer) clearInterval(keepAliveTimer);
  if (server) {
    server.close(() => {
      console.log('Server closed. Exiting process...');
      process.exit(0);
    });
  } else {
    process.exit(0);
  }
});

process.on('SIGTERM', () => {
  console.log('Received SIGTERM. Shutting down AI service gracefully...');
  if (keepAliveTimer) clearInterval(keepAliveTimer);
  if (server) {
    server.close(() => {
      console.log('Server closed. Exiting process...');
      process.exit(0);
    });
  } else {
    process.exit(0);
  }
});

// Prevent uncaught exceptions from crashing the app
process.on('uncaughtException', (error) => {
  console.error('Uncaught Exception:', error);
  console.error(error.stack);
});

// Prevent unhandled promise rejections from crashing the app
process.on('unhandledRejection', (reason, promise) => {
  console.error('Unhandled Promise Rejection:', reason);
});

// Session management functions
function getOrCreateSession(sessionId = null) {
  // Check if we should clean up old sessions
  const now = new Date();
  if ((now - LAST_CLEANUP_TIME) > SESSION_CLEANUP_INTERVAL) {
    cleanupExpiredSessions();
    LAST_CLEANUP_TIME = now;
  }
  
  // If session_id provided and exists, return it
  if (sessionId && sessions[sessionId]) {
    // Update last activity
    sessions[sessionId].last_activity = now;
    return sessionId;
  }
  
  // Otherwise create a new session
  const newSessionId = uuidv4();
  sessions[newSessionId] = {
    created_at: now,
    last_activity: now,
    message_history: [
      {
        role: "system", 
        content: `You are an expert programmer focusing on improving code.
          You provide clear, detailed explanations of your changes.
          You focus on making code more readable, maintainable, efficient,
          and following best practices for the language.`
      }
    ],
    code_history: []
  };
  return newSessionId;
}

function cleanupExpiredSessions() {
  const now = new Date();
  const expiredSessions = [];
  
  for (const sessionId in sessions) {
    if ((now - sessions[sessionId].last_activity) > SESSION_EXPIRY) {
      expiredSessions.push(sessionId);
    }
  }
  
  for (const sessionId of expiredSessions) {
    delete sessions[sessionId];
  }
  
  if (expiredSessions.length) {
    console.log(`Cleaned up ${expiredSessions.length} expired sessions`);
  }
}

// API Endpoints
app.post('/ai/improve', async (req, res) => {
  try {
    const { code, language, prompt, session_id } = req.body;
    
    // Get or create a session
    const sessionId = getOrCreateSession(session_id);
    const session = sessions[sessionId];
    
    console.log(`Processing code improvement: session=${sessionId}, language=${language}, prompt=${prompt}`);
    
    // Add the current code to history
    const codeEntry = {
      timestamp: new Date().toISOString(),
      language,
      code,
      prompt
    };
    session.code_history.push(codeEntry);
    
    // Prepare context from previous interactions
    let context = "";
    if (session.code_history.length > 1) {
      // Include information about previous changes
      const prevEntries = session.code_history.slice(-2);  // Get last 2 entries (excluding current)
      prevEntries.forEach((entry, i) => {
        context += `Previous code version ${i+1}:\n\`\`\`${entry.language}\n${entry.code}\n\`\`\`\n`;
        context += `Previous request: ${entry.prompt}\n\n`;
      });
    }
    
    // Create a prompt for the Groq model
    const systemPrompt = `
      You are an expert programmer focusing on improving code. 
      When asked to improve code, you will:
      1. Maintain the same general functionality unless explicitly asked for changes
      2. Make the code more efficient, readable, or maintainable as appropriate
      3. Add comments to explain complex parts
      4. First provide ONLY the improved code without any explanation or markdown formatting
      5. Then on a new line write "EXPLANATION:" followed by your explanation of the changes
      
      Do not include any additional text before or after these two parts.
    `;
    
    const userPrompt = `
      ${context}
      
      Please improve this ${language} code according to these requirements:
      
      ${prompt}
      
      Here is the code to improve:
      \`\`\`${language}
      ${code}
      \`\`\`
      
      Remember to structure your response with the improved code first, followed by "EXPLANATION:" and then your explanation.
    `;
    
    // Add the user request to the message history
    session.message_history.push({ role: "user", content: userPrompt });
    
    // Keep message history to a reasonable size (last 10 messages)
    if (session.message_history.length > 12) {  // system + 10 exchanges + current
      session.message_history = [session.message_history[0]].concat(session.message_history.slice(-11));
    }
    
    console.log("Sending request to Groq API with message history...");
    const response = await groqClient.chat.completions.create({
      model: MODEL_NAME,
      messages: session.message_history,
      temperature: 0.2,
      max_tokens: 4000,
    });
    
    const result = response.choices[0].message.content;
    console.log(`Received response from Groq API. Length: ${result.length} characters`);
    if (response.usage.completion_time) {
      console.log(`Response completion time: ${response.usage.completion_time}s`);
    }
    
    // Add assistant response to history
    session.message_history.push({ role: "assistant", content: result });
    
    // Parse the result to separate code and explanation
    let parts = result.split("EXPLANATION:");
    let updatedCode, explanation;
    
    if (parts.length > 1) {
      updatedCode = parts[0].trim();
      explanation = parts[1].trim();
    } else {
      // If no EXPLANATION section, try to guess the separation
      const lines = result.split('\n');
      let codeEndIndex = -1;
      
      // Look for markdown code blocks or natural language transitions
      for (let i = 0; i < lines.length; i++) {
        if (lines[i].includes("```") || lines[i].includes("Here's why:") || lines[i].includes("Explanation:")) {
          codeEndIndex = i;
          break;
        }
      }
      
      if (codeEndIndex > 0) {
        updatedCode = lines.slice(0, codeEndIndex).join('\n').trim();
        explanation = lines.slice(codeEndIndex).join('\n').trim();
      } else {
        // Fallback: assume 80% is code, rest is explanation
        const splitPoint = Math.floor(lines.length * 0.8);
        updatedCode = lines.slice(0, splitPoint).join('\n').trim();
        explanation = lines.slice(splitPoint).join('\n').trim();
      }
    }
    
    // Clean up the code (remove markdown code blocks if present)
    if (updatedCode.startsWith("```")) {
      const codeLines = updatedCode.split('\n');
      // Remove first line with ``` and potentially the language identifier after it
      if (codeLines.length >= 3 && codeLines[0].includes("```")) {
        // Determine where the end of the code block is
        let endIndex = null;
        for (let i = 1; i < codeLines.length; i++) {
          if (codeLines[i].includes("```")) {
            endIndex = i;
            break;
          }
        }
        
        if (endIndex) {
          updatedCode = codeLines.slice(1, endIndex).join('\n');
        } else {
          // No closing marker found, just remove the first line
          updatedCode = codeLines.slice(1).join('\n');
        }
      }
    }
    
    console.log("Successfully processed code improvement request");
    return res.json({
      updated_code: updatedCode,
      explanation: explanation,
      session_id: sessionId
    });
  } catch (e) {
    console.error(`AI processing error: ${e.message}`);
    console.error(e.stack);
    return res.status(500).json({ detail: `AI processing error: ${e.message}` });
  }
});

// Session management endpoints
app.get('/ai/sessions', (req, res) => {
  const sessionList = [];
  for (const sessionId in sessions) {
    sessionList.push({
      session_id: sessionId,
      created_at: sessions[sessionId].created_at.toISOString(),
      last_activity: sessions[sessionId].last_activity.toISOString(),
      message_count: sessions[sessionId].message_history.length,
      code_versions: sessions[sessionId].code_history.length
    });
  }
  return res.json(sessionList);
});

app.get('/ai/session/:sessionId', (req, res) => {
  const { sessionId } = req.params;
  if (!sessions[sessionId]) {
    return res.status(404).json({ detail: "Session not found" });
  }
  
  const session = sessions[sessionId];
  return res.json({
    session_id: sessionId,
    created_at: session.created_at.toISOString(),
    last_activity: session.last_activity.toISOString(),
    message_count: session.message_history.length,
    code_versions: session.code_history.length
  });
});

app.get('/ai/session/:sessionId/history', (req, res) => {
  const { sessionId } = req.params;
  if (!sessions[sessionId]) {
    return res.status(404).json({ detail: "Session not found" });
  }
  
  const session = sessions[sessionId];
  return res.json({
    session_id: sessionId,
    message_history: session.message_history,
    code_history: session.code_history
  });
});

app.delete('/ai/session/:sessionId', (req, res) => {
  const { sessionId } = req.params;
  if (!sessions[sessionId]) {
    return res.status(404).json({ detail: "Session not found" });
  }
  
  delete sessions[sessionId];
  return res.json({ message: `Session ${sessionId} deleted successfully` });
});

app.get('/test-groq', async (req, res) => {
  try {
    const response = await groqClient.chat.completions.create({
      model: MODEL_NAME,
      messages: [
        { role: "system", content: "You are a helpful assistant." },
        { role: "user", content: "Explain what makes Groq special in 2 sentences." }
      ],
      temperature: 0.2,
      max_tokens: 100
    });
    
    return res.json({
      status: "success",
      message: "Groq API is working",
      response: response.choices[0].message.content,
      completion_time: response.usage?.completion_time,
      model: MODEL_NAME
    });
  } catch (e) {
    return res.json({
      status: "error",
      message: `Failed to connect to Groq API: ${e.message}`,
      traceback: e.stack
    });
  }
});

app.get('/ping', (req, res) => {
  return res.json({ message: "pong" });
});

// Start the server only if this file is being run directly
if (import.meta.url === `file://${process.argv[1]}` || process.argv[1] === undefined) {
  // Run as an IIFE (Immediately Invoked Function Expression) to use async/await
  (async () => {
    try {
      // Test the Groq API
      await testGroqApi();
      
      // Create an HTTP server explicitly (more reliable than app.listen)
      const PORT = process.env.AI_PORT || 8001;
      server = http.createServer(app);
      
      // Add error handling for the server
      server.on('error', (error) => {
        console.error(`Server error: ${error.message}`);
        console.error(error.stack);
      });
      
      // Start the server
      server.listen(PORT, '0.0.0.0', () => {
        console.log(`AI service running on port ${PORT}`);
        console.log(`Server started at: ${new Date().toLocaleString()}`);
        console.log(`Press Ctrl+C to stop the service`);
        
        // Start the keep-alive mechanism to prevent the process from exiting
        startKeepAlive();
      });
      
      // This is important: create a permanent reference to prevent garbage collection
      global.server = server;
      
    } catch (error) {
      console.error(`Failed to start AI service: ${error.message}`);
      console.error(error.stack);
      process.exit(1);
    }
  })().catch(error => {
    console.error(`Unhandled startup error: ${error.message}`);
    console.error(error.stack);
    process.exit(1);
  });

  // Create a non-clearing interval to keep the Node.js process running
  setInterval(() => {
    // This empty function keeps the event loop active
  }, 1000 * 60 * 60); // Run every hour
}

export default app;